drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/rulesengine/MemoryMonitorUtil.java

```diff
@@ -0,0 +1,46 @@
+package org.drools.ansible.rulebook.integration.api.rulesengine;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MemoryMonitorUtil {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MemoryMonitorUtil.class.getName());
+
+    public static final String MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY = "drools.memory.occupation.percentage.threshold";
+
+    private static final int DEFAULT_MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD = 90;
+
+    private static final int MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD;
+
+    static {
+        String envValue = System.getenv("DROOLS_MEMORY_THRESHOLD");
+        if (envValue != null && !envValue.isEmpty()) {
+            // Environment variable takes precedence over system property
+            System.setProperty(MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY, envValue);
+        }
+        MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD = Integer.getInteger(MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD_PROPERTY, DEFAULT_MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD); // percentage
+    }
+
+    private MemoryMonitorUtil() {
+        // do not instantiate
+    }
+
+    public static void checkMemoryOccupation() {
+        int memoryOccupationPercentage = getMemoryOccupationPercentage();
+        if (memoryOccupationPercentage > MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD) {
+            System.gc(); // NOSONAR
+            // double check to avoid frequent GC
+            memoryOccupationPercentage = getMemoryOccupationPercentage();
+            if (memoryOccupationPercentage > MEMORY_OCCUPATION_PERCENTAGE_THRESHOLD) {
```

Comment:
This means I call `System.gc()` only when heap exceeds the threshold. However, I think it doesn't make much sense because GC is processed by a different thread (and not something we can fully control) so the second check wouldn't reflect the result of GC. Shall I simply remove it?

Also Shall I add another threshold (e.g. 60%) to call `System.gc()`?

As far as I observe the default G1GC, GC is triggered when the used heap is around 60% - 75% of the total heap.

Please let me know your thoughts, @mariofusco

Reply:
I believe that what you did here is a good idea. Let's give the gc one last chance to run before actually reporting the problem.

Reply:
Thanks, I keep the logic.